web/web.go

```diff
@@ -0,0 +1,1022 @@
+	// version of 'toDDL' with the new type mapping).
+	for t, spSchema := range app.conv.SpSchema {
+		for col, _ := range spSchema.ColDefs {
+			sourceTable := app.conv.ToSource[t].Name
+			sourceCol := app.conv.ToSource[t].Cols[col]
+			srcCol := app.conv.SrcSchema[sourceTable].ColDefs[sourceCol]
+			// If the srcCol's type is in the map, then recalculate the Spanner type
+			// for this column using the map. Otherwise, leave the ColDef for this
+			// column as is. Note that per-column type overrides could be lost in
+			// this process -- the mapping in typeMap always takes precendence.
+			if _, found := typeMap[srcCol.Type.Name]; found {
+				updateType(typeMap[srcCol.Type.Name], t, col, sourceTable, w)
+			}
+		}
+	}
+	updateSessionFile()
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(app.conv)
+}
+
+// Actions to be performed on a column.
+// (1) Removed: true/false
+// (2) Rename: New name or empty string
+// (3) PK: "ADDED", "REMOVED" or ""
+// (4) NotNull: "ADDED", "REMOVED" or ""
+// (5) ToType: New type or empty string
+type updateCol struct {
+	Removed bool   `json:"Removed"`
+	Rename  string `json:"Rename"`
+	PK      string `json:"PK"`
+	NotNull string `json:"NotNull"`
+	ToType  string `json:"ToType"`
+}
+
+type updateTable struct {
+	UpdateCols map[string]updateCol `json:"UpdateCols"`
+}
+
+// updateTableSchema updates the Spanner schema.
+// Following actions can be performed on a specified table:
+// (1) Remove column
+// (2) Rename column
+// (3) Add or Remove Primary Key
+// (4) Add or Remove NotNull constraint
+// (5) Update Spanner type
+func updateTableSchema(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var t updateTable
+	table := r.FormValue("table")
+	err = json.Unmarshal(reqBody, &t)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	srcTableName := app.conv.ToSource[table].Name
+	for colName, v := range t.UpdateCols {
+		if v.Removed {
+			err, status := canRemoveColumn(colName, table)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("%v", err), status)
+				return
+			}
+			removeColumn(table, colName, srcTableName)
+			continue
+		}
+		newColName := colName
+		if v.Rename != "" && v.Rename != colName {
+			err, status := canRenameOrChangeType(colName, table)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("%v", err), status)
+				return
+			}
+			renameColumn(v.Rename, table, colName, srcTableName)
+			newColName = v.Rename
+		}
+		if v.PK != "" {
+			// Note: Adding or removing PK feature is deferred at the moment.
+
+			// err, status := canChangePK(v.PK, table)
+			// if err != nil {
+			// 	http.Error(w, fmt.Sprintf("%v", err), status)
+			// 	return
+			// }
+			// pkChanged(v.PK, table, newColName)
+		}
```

Comment:
I would suggest deleting this commented code and simply return an error saying that "HarbourBridge currently doesn't support editing primary keys". We should then also remove the corresponding function pkChanged.

Reply:
Yes agreed.
It is important that we return an error here so the UI shows an appropriate message.

Reply:
done